### PR TITLE
x/sqlbuilder: Initial implementation of an SQL query builder

### PR DESCRIPTION
**Pull request by AlexisMontagne**

### CODE REVIEW

File: x/sqlbuilder/predicate_clause.go [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+type plainSQLPredicate string
+
+func (psp plainSQLPredicate) ToSQL() string { return string(psp) }
+
+func PlainSQLPredicate(exp string) PredicateClause {
+	return &staticStmtPredicateClauseWrapper{sspc: plainSQLPredicate(exp)}
+}
+
+func EqMarkers(l, r Marker) PredicateClause {
+	return &staticStmtPredicateClauseWrapper{
```
Comment (tgallice) [inline x/sqlbuilder/predicate_clause.go]: you can do `return PlainSQLPredicate(fmt.Sprintf("%s = %s", l.ToSQL(), r.ToSQL()))`

File: x/sqlbuilder/select_statement.go [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+	SelectClauses []Marker
+	WhereClause   PredicateClause
+	GroupByClause []Marker
+	HavingClause  PredicateClause
+}
+
+type JoinType string
+
+const (
+	DefaultJoin JoinType = ""
```
Comment (tgallice) [inline x/sqlbuilder/select_statement.go:22]: What is the use of this ?

Comment (AlexisMontagne) [reply x/sqlbuilder/select_statement.go:22]: JOIN with no INNER nor OUTER  token in front of it, is a valid ANSI SQL expression. therefore i tried to represent it here, may be useless? on tje DB systems with use it is equivalent to a INNER JOIN

File: x/sqlbuilder/query_builder.go [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+
+func (qb *QueryBuilder) PrepareUpdate(us UpdateStatement) Execer {
+	return &execer{qb: qb, stmt: us}
+}
+
+func (qb *QueryBuilder) PrepareDelete(ds DeleteStatement) Execer {
+	return &execer{qb: qb, stmt: ds}
+}
+
+type Execer interface {
```
Comment (Sypheos) [inline x/sqlbuilder/query_builder.go:32]: Do you mean `Executer` or `Executor` ?

File: x/sqlbuilder/query_builder.go [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+
+type Cursor interface {
+	Scanner
+
+	Close() error
+	Err() error
+	Next() bool
+}
+
+type Queryer interface {
```
Comment (Sypheos) [inline x/sqlbuilder/query_builder.go:67]: `Querier`

File: x/sqlbuilder/select_statement.go [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+	return jc.WhereClause.WriteTo(w, vs)
+}
+
+type QueryWriter interface {
+	io.Writer
+
+	RedeemVariable(interface{}) string
+}
+
+type queryWriter struct {
```
Comment (Sypheos) [inline x/sqlbuilder/select_statement.go]: Looks like it is used in other statement, maybe move it to `query_builder`?

Also it seems weird to have a `writer` and a `builder` both terms are semantically close and `writer` seems to have the same role as a `strings.Builder`.

**Review by Sypheos — COMMENTED** [review]
Kinda confused about the naming but OK.

File: x/sqlbuilder/query_builder.go (hunk 3 above)
Comment (AlexisMontagne) [reply x/sqlbuilder/query_builder.go:32]: I tried to keep a semantic close to the stdlib one, kinda bugged me too at first but: https://golang.org/src/database/sql/driver/driver.go?s=5253:5328#L137

File: x/sqlbuilder/query_builder.go (hunk 4 above)
Comment (AlexisMontagne) [reply x/sqlbuilder/query_builder.go:67]: Same reason as above https://golang.org/src/database/sql/driver/driver.go?s=6249:6324#L164

File: x/sqlbuilder/select_statement.go (hunk 5 above)
Comment (AlexisMontagne) [reply x/sqlbuilder/select_statement.go]: > Looks like it is used in other statement, maybe move it to query_builder?

 you are totally right, ill do it

> Also it seems weird to have a writer and a builder both terms are semantically close and writer seems to have the same role as a strings.Builder. 

The idea behind the semantic is the following:

The `QueryBuilder` takes some state-less building block called `*Statement`. In order to build a query ( statement + arguments) a `*Statement` and his subsequent structs know how to write down SQL bits on to the current state of the query. This current state of the query is held by the `QueryWriter` implementation. It follows the semantic of the stdlib's io package

**Review by tgallice — APPROVED** [review]

File: x/sqlbuilder/query_builder.go (hunk 4 above)
Comment (Sypheos) [reply x/sqlbuilder/query_builder.go:67]: :man_facepalming: ok

**Review by Sypheos — APPROVED** [review]

**Pull request merged**